src/Access/LDAPAccessStorage.cpp

```diff
@@ -0,0 +1,197 @@
+
+    const bool has_server = config.has(prefix_str + "server");
+    const bool has_user_template = config.has(prefix_str + "user_template");
+
+    if (!has_server)
+        throw Exception("Missing 'server' field for LDAP user directory.", ErrorCodes::BAD_ARGUMENTS);
+
+    const auto ldap_server_cfg = config.getString(prefix_str + "server");
+    String user_template_cfg;
+
+    if (ldap_server_cfg.empty())
+        throw Exception("Empty 'server' field for LDAP user directory.", ErrorCodes::BAD_ARGUMENTS);
+
+    if (has_user_template)
+        user_template_cfg = config.getString(prefix_str + "user_template");
+
+    if (user_template_cfg.empty())
+        user_template_cfg = "default";
+
+    ldap_server = ldap_server_cfg;
+    user_template = user_template_cfg;
+    top_enclosing_storage = top_enclosing_storage_;
+}
+
+
+bool LDAPAccessStorage::isConfiguredNoLock() const
+{
+    return !ldap_server.empty() && !user_template.empty() && top_enclosing_storage;
+}
+
+
+const char * LDAPAccessStorage::getStorageType() const
+{
+    return STORAGE_TYPE;
+}
+
+
+bool LDAPAccessStorage::isStorageReadOnly() const
+{
+    return true;
+}
+
+
+std::optional<UUID> LDAPAccessStorage::findImpl(EntityType type, const String & name) const
+{
+    if (type == EntityType::USER)
+    {
+        std::scoped_lock lock(mutex);
+
+        // Detect and avoid loops/duplicate creations.
+        if (helper_lookup_in_progress)
+            return {};
+
+        helper_lookup_in_progress = true;
+        SCOPE_EXIT({ helper_lookup_in_progress = false; });
+
+        // Return the id immediately if we already have it.
+        const auto id = memory_storage.find(type, name);
+        if (id.has_value())
+            return id;
+
+        if (!isConfiguredNoLock())
+            return {};
+
+        // Stop if entity exists anywhere else, to avoid duplicates.
+        if (top_enclosing_storage->find(type, name))
+            return {};
+
+        // Entity doesn't exist. We are going to create one.
+
+        // Retrieve the template first.
+        std::shared_ptr<const User> user_tmp;
+        try
+        {
+            user_tmp = top_enclosing_storage->read<User>(user_template);
+            if (!user_tmp)
+                throw Exception("Retrieved user is empty", IAccessEntity::TypeInfo::get(IAccessEntity::Type::USER).not_found_error_code);
+        }
+        catch (...)
+        {
+            tryLogCurrentException(getLogger(), "Unable to retrieve user template '" + user_template + "' from access storage '" + top_enclosing_storage->getStorageName() + "'");
+            return {};
+        }
+
+        // Build the new entity based on the existing template.
+        const auto user = std::make_shared<User>(*user_tmp);
+        user->setName(name);
+        user->authentication = Authentication(Authentication::Type::LDAP_SERVER);
+        user->authentication.setServerName(ldap_server);
+
+        return memory_storage.insert(user);
```

Comment:
I think cloning the user should be implemented in more thoughtful way. I mean there are access rights associated with the user template and if they're changed we should change all clones too. Another issue about the cloning is that there are quotas and row policies and they keep IDs of users they're attached to.

Reply:
The last issue doesn't seem to be easy to solve, so I suppose for the first implementation we can just ignore that.

Reply:
> I mean there are access rights associated with the user template and if they're changed we should change all clones too.

They are recreated each time, and the up-to-date template user is taken, or you mean something else?

Reply:
I mean what if the template is changed after you create one of those clones? You don't change clones after they're created.

Reply:
How long are `User` kept around in caches? I create a clone when I need to try to authenticate a used, and if auth attempt succeeds, that `User` will be used further in the session. So if the object is not cached, a clone exists until deleted, and the next time the same user tries to authenticate a `User` object will be recreated using possibly new template user.

Reply:
It would be logical to only "inherit" roles from the `user_template` and that way any updates to the roles will be propagated to all the users (by design). This would work for privileges, quotas, row policies and settings profiles.

Reply:
> It would be logical to only "inherit" roles from the `user_template` and that way any updates to the roles will be propagated to all the users (by design). This would work for privileges, quotas, row policies and settings profiles.

Great idea! `LDAPAccessStorage` can read role names instead of the name of a user template from the main config and then create all users with all those role instead of cloning. It will solve the issue with updating as you said.

Reply:
Changed to that.